### PR TITLE
[Data][Docs] Consolidate Data user guides

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

The Data documentation contains obsolete or unnecessary information. In addition to general revisions, this PR:
* Removes "FAQ"
* Moves advanced content in "Key concepts" to "Scheduling, Memory, and Execution"
* Merges "Getting started" and "Key concepts"
* Renames "Scheduling, Memory, and Execution" to "Data internals"

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/data/data-internals.rst [hunk 1]
```diff
@@ -1,42 +1,79 @@
 .. _datasets_scheduling:
 
-============================================
-Scheduling, Execution, and Memory Management
-============================================
+==================
+Ray Data Internals
+==================
```
Comment (amogkam) [inline doc/source/data/data-internals.rst:6]: add a sentence on what this page is about?

File: doc/source/data/data-internals.rst [hunk 2]
```diff
@@ -1,42 +1,79 @@
...
+Operations
 ==========
 
-Ray Data uses Ray core for execution, and hence is subject to the same scheduling considerations as normal Ray tasks and actors. Ray Data uses the following custom scheduling settings by default for improved performance:
+Reading files
+-------------
 
-* The ``SPREAD`` scheduling strategy is used to ensure data blocks are evenly balanced across the cluster.
-* Retries of application-level exceptions are enabled to handle transient errors from remote datasources.
-* Dataset tasks ignore placement groups by default, see :ref:`Ray Data and Placement Groups <datasets_pg>`.
```
Comment (amogkam) [inline doc/source/data/data-internals.rst:14]: This content is still relevant to keep in a "Dataset internals" page?

File: doc/source/data/data-internals.rst [hunk 3]
```diff
@@ -1,42 +1,79 @@
...
+    words, shuffle execution isn't streamed through memory.
+
+For an in-depth guide on shuffle performance, see :ref:`Performance Tips and Tuning <shuffle_performance_tips>`.
+
+Scheduling
+==========
+
+Ray Data uses Ray core for execution, and hence is subject to the same scheduling considerations as normal Ray tasks and actors. Ray Data uses the following custom scheduling settings by default for improved performance:
+
+* The ``SPREAD`` scheduling strategy is used to ensure data blocks are evenly balanced across the cluster.
```
Comment (amogkam) [inline doc/source/data/data-internals.rst:74]: Also to ensure that map tasks are balanced across the cluster as well.

File: doc/source/data/data-internals.rst [hunk 4]
```diff
@@ -1,42 +1,79 @@
...
+* Retries of application-level exceptions are enabled to handle transient errors from remote datasources.
```
Comment (amogkam) [inline doc/source/data/data-internals.rst:75]: this sentence only applies to read stages, right? And it's not really about scheduling. I think we should move this detail to the reading section.

File: doc/source/data/data-internals.rst [hunk 5]
```diff
@@ -1,42 +1,79 @@
...
+* Dataset tasks ignore placement groups by default, see :ref:`Ray Data and Placement Groups <datasets_pg>`.
```
Comment (amogkam) [inline doc/source/data/data-internals.rst:76]: Did we still want to keep the additional details about Ray Data interaction with placement groups and Tune?

File: doc/source/data/overview.rst [hunk 6]
```diff
@@ -23,6 +23,10 @@ It provides flexible and performant APIs for distributed data processing:
 
 Ray Data is built on top of Ray, so it scales effectively to large clusters and offers scheduling support for both CPU and GPU resources. Ray Data uses `streaming execution <https://www.anyscale.com/blog/streaming-distributed-execution-across-cpus-and-gpus>`__ to efficiently process large datasets.
 
+.. note::
+
+    Ray Data isn't a replacement for generic ETL pipelines like Spark or distributed
+    DataFrame libraries like Dask.
```
Comment (amogkam) [inline doc/source/data/overview.rst:29]: I would reword this a bit with our most recent positioning. 

"Ray Data does not have a SQL interface and is not meant as a replacement for generic ETL pipelines like Spark."

I would remove the mention of Dask.

File: doc/source/data/performance-tips.rst [hunk 7]
```diff
@@ -3,105 +3,40 @@
...
-    * Batch iteration time breakdown (summed across prefetch threads):
-        * In ray.get(): 75.68us min, 220.26us max, 131.89us avg, 2.11ms total
-        * In batch creation: 326.58us min, 1.37ms max, 644.86us avg, 25.79ms total
-        * In batch formatting: 101.81us min, 898.73us max, 172.38us avg, 6.9ms total
-
-Batching Transforms
+Batching transforms
 ~~~~~~~~~~~~~~~~~~~
 
 Mapping individual records using :meth:`.map(fn) <ray.data.Dataset.map>` can be quite slow.
```
Comment (amogkam) [inline doc/source/data/performance-tips.rst:12]: In what cases can it be slow?

File: doc/source/data/performance-tips.rst [hunk 8]
```diff
@@ -3,105 +3,40 @@
...
-Instead, consider using :meth:`.map_batches(batch_fn, batch_format="pandas") <ray.data.Dataset.map_batches>` and writing your ``batch_fn`` to
-perform vectorized pandas operations.
-
-.. _data_format_overheads:
+Instead, consider using :meth:`.map_batches(batch_fn) <ray.data.Dataset.map_batches>` and writing your ``batch_fn`` to
+perform vectorized operations.
```
Comment (amogkam) [inline doc/source/data/performance-tips.rst:14]: I think we want to clarify this more, otherwise it ends up with the same confusion as the warning.

I think let's explicitly say there is no performance benefit if your UDF is not vectorized.

File: doc/source/data/performance-tips.rst [hunk 9]
```diff
@@ -3,105 +3,40 @@
...
 
+By default, Ray Data automatically selects the read ``parallelism`` according to the following procedure:
 
-.. list-table:: Data Format Conversion Costs
-   :header-rows: 1
-   :stub-columns: 1
+1. The number of available CPUs is estimated. If in a placement group, the number of CPUs in the cluster is scaled by the size of the placement group compared to the cluster size. If not in a placement group, this is the number of CPUs in the cluster.
+2. The parallelism is set to the estimated number of CPUs multiplied by 2. If the parallelism is less than 8, it is set to 8.
+3. The in-memory data size is estimated. If the parallelism would create in-memory blocks that are larger on average than the target block size (512MiB), the parallelism is increased until the blocks are < 512MiB in size.
+4. The parallelism is truncated to ``min(num_files, parallelism)``.
```
Comment (amogkam) [inline doc/source/data/performance-tips.rst:27]: This step is no longer applied with the SplitBlocks operator. @ericl to confirm

File: doc/source/data/performance-tips.rst [hunk 10]
```diff
@@ -189,3 +134,79 @@ setting the ``DataContext.use_push_based_shuffle`` flag:
...
+
+   # By default, this is set to False.
+   ctx.execution_options.preserve_order = True
+
+To enable deterministic execution, set the above to True. This may decrease performance, but will ensure block ordering is preserved through execution. This flag defaults to False.
+
+.. _datasets_pg:
+
+Ray Data and placement groups
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (amogkam) [inline doc/source/data/performance-tips.rst:186]: Does this belong under reproducibility? It feels more like scheduling details and less of a performance tip

File: doc/source/data/performance-tips.rst [hunk 11]
```diff
@@ -189,3 +134,79 @@ setting the ``DataContext.use_push_based_shuffle`` flag:
...
+By default, Ray Data configures its tasks and actors to use the cluster-default scheduling strategy ("DEFAULT"). You can inspect this configuration variable here:
+:class:`ray.data.DataContext.get_current().scheduling_strategy <ray.data.DataContext>`. This scheduling strategy will schedule these tasks and actors outside any present
+placement group. If you want to force Ray Data to schedule tasks within the current placement group (i.e., to use current placement group resources specifically for Ray Data), you can set ``ray.data.DataContext.get_current().scheduling_strategy = None``.
+
+This should be considered for advanced use cases to improve performance predictability only. We generally recommend letting Ray Data run outside placement groups as documented in the :ref:`Ray Data and Other Libraries <datasets_tune>` section.
+
+.. _datasets_tune:
+
+Ray Data and Tune
+-----------------
```
Comment (amogkam) [inline doc/source/data/performance-tips.rst:197]: Same thing here

**Review by amogkam — COMMENTED** [review]
nice work! Completely agree a lot of the current information in Key Concepts is too low level. Thanks for moving this into Dataset internals.

But I still think we want a Key concepts that describes the basics and for consistency with other libraries. Given there's a lot of overlap with "Getting Started", should we consolidate the two pages, similar to Ray Serve Key concepts?

File: doc/source/data/data-internals.rst [hunk 12]
```diff
@@ -1,22 +1,102 @@
...
+.. _datasets_pg:
+
+Ray Data and placement groups
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+By default, Ray Data configures its tasks and actors to use the cluster-default scheduling strategy ("DEFAULT"). You can inspect this configuration variable here:
+:class:`ray.data.DataContext.get_current().scheduling_strategy <ray.data.DataContext>`. This scheduling strategy will schedule these tasks and actors outside any present
+placement group. If you want to force Ray Data to schedule tasks within the current placement group (i.e., to use current placement group resources specifically for Ray Data), you can set ``ray.data.DataContext.get_current().scheduling_strategy = None``.
+
+This should be considered for advanced use cases to improve performance predictability only. We generally recommend letting Ray Data run outside placement groups as documented in the :ref:`Ray Data and Other Libraries <datasets_tune>` section.
```
Comment (amogkam) [inline doc/source/data/data-internals.rst:94]: I don't think the reference here makes sense anymore. Let's just remove it?

File: doc/source/data/data.rst [hunk 13]
```diff
@@ -66,22 +78,6 @@ Learn more
 
             Learn how to use Ray Data
 
-    .. grid-item-card::
-
-        **Key Concepts**
```
Comment (amogkam) [inline doc/source/data/data.rst:71]: Can we update the description in the Gettin started card now that it is more of a Key Concepts page?

File: doc/source/ray-air/examples/dolly_lightning_fsdp_finetuning.ipynb [hunk 14]
```diff
@@ -139,6 +139,7 @@
    ]
   },
   {
+   "attachments": {},
```
Comment (amogkam) [inline doc/source/ray-air/examples/dolly_lightning_fsdp_finetuning.ipynb:142]: revert these changes?

File: doc/source/data/key-concepts.rst [hunk 15]
```diff
@@ -1,107 +1,149 @@
...
-.. _dataset_concept:
+Datasets
+--------
 
-----------
-Dataset
-----------
+Ray Data's main abstraction is a :class:`Dataset <ray.data.Dataset>`, which
+is a distributed data transformation pipeline. Ray Data provides APIs for loading
+data and streaming transformations in parallel.
```
Comment (amogkam) [inline doc/source/data/key-concepts.rst:11]: I think we want to reword this. Something like below, with the appropriate links

Ray Data's main abstraction is a Dataset which is a distributed collection of data.

Ray Data provides APIs for loading large amounts of data, transforming data in parallel, and consuming data.

**Review by amogkam — COMMENTED** [review]
some minor comments. overall looks good

**Review by matthewdeng — APPROVED** [review]
Stamp for Tune.

**Review by amogkam — APPROVED** [review]
nice work!

**Comment by amogkam** [comment]
there are some failing tests

**Review by angelinalg — APPROVED** [review]
